common/OpTestSystem.py

```diff
@@ -141,6 +147,13 @@ def set_state(self, state):
         self.state = state
 
     def goto_state(self, state):
+        # only perform detection when incoming state is UNKNOWN
+        # if user overrides from command line and machine not at desired state can lead to exceptions
+        if (self.state == OpSystemState.UNKNOWN):
+          print "OpTestSystem CHECKING CURRENT STATE and TRANSITIONING for TARGET STATE: %s" % (state)
+          self.state = self.run_DETECT(state)
+          print "OpTestSystem CURRENT DETECTED STATE: %s" % (self.state)
+
         print "OpTestSystem START STATE: %s (target %s)" % (self.state, state)
         never_unknown = False
         while 1:
```

Comment:
I think in this while loop, we need to consider not to go into state UNKNOWN_BAD,

Reply:
> state UNKNOWN_BAD

The only time a transition from UNKNOWN_BAD will occur will be during an active test when self.state will be set to UNKNOWN_BAD, the UNKNOWN self.state is only used as a "first time" flag indicator in the while loop.  So if UNKNOWN_BAD comes into the while loop it is handled just as any other case, and then jumps to the state machine run_UNKNOWN which is the proper transition, at that time the UNKNOWN_BAD state "disappears" and allows it to transition just as prior to this patch.

Reply:
> UNKNOWN_BAD

UNKNOWN_BAD is not used as a state transition during the active state machine transition stages.

Reply:
Yeah, i am considering a case here when state machine goes through different states, it shouldn't end up in going from a good state to UNKNOWN_BAD. i think as of now this is fine, that case well never happen. so no worries.

Reply:
```
        if never_unknown and self.state == OpSystemState.UNKNOWN:
                raise 'System State transition failure: should not have progressed to UNKNOWN!'
```
The above condition will check for not going to UNKNOWN state from a known good state(like IPLing, OS etc.)